lib/class-wp-rest-search-controller.php

```diff
@@ -0,0 +1,539 @@
+			'items'       => array(
+				'enum' => array_merge( array_keys( $this->get_allowed_post_types() ), array( 'any' ) ),
+				'type' => 'string',
+			),
+		);
+
+		$query_params['type_exclude'] = array(
+			'default'     => array(),
+			'description' => __( 'Ensure search results exclude content of one or more specific types.', 'gutenberg' ),
+			'type'        => 'array',
+			'items'       => array(
+				'enum' => array_keys( $this->get_allowed_post_types() ),
+				'type' => 'string',
+			),
+		);
+
+		return $query_params;
+	}
+
+	/**
+	 * Prepares links for the request.
+	 *
+	 * @since 2.9.0
+	 *
+	 * @param WP_Post $post Post object.
+	 * @return array Links for the given post.
+	 */
+	protected function prepare_links( $post ) {
+
+		$links = array();
+
+		$item_route = $this->detect_rest_item_route( $post );
+		if ( ! empty( $item_route ) ) {
+			$links['self'] = array(
+				'href'       => rest_url( $item_route ),
+				'embeddable' => true,
+			);
+		}
+
+		$links['collection'] = array(
+			'href' => rest_url( sprintf( '%s/%s', $this->namespace, $this->rest_base ) ),
+		);
+
+		$links['about'] = array(
+			'href' => rest_url( 'wp/v2/types/' . $post->post_type ),
+		);
+
+		if ( ! empty( $post->post_author ) ) {
+			$links['author'] = array(
+				'href'       => rest_url( 'wp/v2/users/' . $post->post_author ),
+				'embeddable' => true,
+			);
+		}
+
+		return $links;
+	}
+
+	/**
+	 * Attempts to detect the route to access a single item.
+	 *
+	 * @since 2.9.0
+	 *
+	 * @param WP_Post $post Post object.
+	 * @return string REST route relative to the REST base URI, or empty string if not found.
+	 */
+	protected function detect_rest_item_route( $post ) {
+		$post_type = get_post_type_object( $post->post_type );
+		if ( ! $post_type ) {
+			return '';
+		}
+
+		// It's currently impossible to detect the REST URL from a custom controller.
+		if ( ! empty( $post_type->rest_controller_class ) && 'WP_REST_Posts_Controller' !== $post_type->rest_controller_class ) {
+			return '';
+		}
+
+		$namespace = 'wp/v2';
+		$rest_base = ! empty( $post_type->rest_base ) ? $post_type->rest_base : $post_type->name;
+
+		return sprintf( '%s/%s/%d', $namespace, $rest_base, $post->ID );
+	}
+
+	/**
+	 * Gets the post statuses allowed for search.
+	 *
+	 * @since 2.9.0
+	 *
+	 * @return array List of post status objects, keyed by their name.
+	 */
+	protected function get_allowed_post_stati() {
+
+		$post_stati = get_post_stati( array(
+			'public'   => true,
+			'internal' => false,
+		), 'objects' );
+
+		$post_stati['inherit'] = get_post_status_object( 'inherit' );
+
+		return $post_stati;
+	}
+
+	/**
+	 * Gets the post types allowed for search.
+	 *
+	 * @since 2.9.0
+	 *
+	 * @param bool $editable Optional. Whether to only include post types where the current user
+	 *                       has edit access. Default false.
+	 * @return array List of post type objects, keyed by their name.
+	 */
+	protected function get_allowed_post_types( $editable = false ) {
```

Comment:
I don't love this `$editable` argument. For the sake of clarity, could we have two separate methods? Or is there a different approach we can take?

Reply:
Makes sense. I will outsource the editing part into another utility method.